File: course_material/Exercises_matlab/Exercise_2/FcnL_phi.m
function L_Phi = FcnL_phi(AK,K,P,alpha)
    % nonlinear system
    mu = 0.5; 
    dynamics = @(x,u) [x(2)+ u*(mu +(1-mu)*x(1)); x(1) + u*(mu -4*(1-mu)*x(2))]; 
    
    % Phi
    phi = @(x) dynamics(x,K*x) - AK*x;
    %phi = @(x)  - AK*x;
  
    % L_Phi
    opt = optimset('MaxFunEvals',10000,'MaxIter',10000,'Display','off');
    
    [x,L_Phi_tilde] = fmincon(@(x) -sqrt(phi(x)' * phi(x))/sqrt(x'*x) ,...
        [10.0;10.0],[],[],[],[],[],[],@(x)nonlinConsAlpha(x,P,alpha),opt);
    
    L_Phi = -L_Phi_tilde;
    

end

function [c, ceq] = nonlinConsAlpha(x, P, alpha)
% All states inside ellipse X_{\alpha}^f = {x \neq 0 | x'*P*x \leq alpha}

    c = x'*P*x - alpha;
    ceq = [];

end

File: course_material/ex_solution/Ex2/FcnL_phi.py
import numpy as np
from scipy.optimize import minimize ,NonlinearConstraint, LinearConstraint




def FcnL_phi_func(AK, K, P, alpha):
    # Nonlinear system
    mu = 0.5 
 
    def lqr(x):
        return K@x

    def dynamics(x):
        mu = 0.5 
        return  np.array([x[1] + lqr(x) * (mu + (1-mu) * x[0]), x[0] + lqr(x) * (mu - 4*(1-mu)*x[1])])    
    

    phi = lambda x: [dynamics(x) - np.dot(AK,x).reshape(-1, 1)] 
    
    
    def nonlinConsAlpha(x):
        # All states inside ellipse X_{alpha}^f = {x \neq 0 | x'*P*x \leq alpha}
        c = x.T @ P @ x - alpha
        return c


    cons_func = lambda x: nonlinConsAlpha(x)
    obj_func = lambda x: - np.linalg.norm(phi(x)) / np.linalg.norm(x)

    

    res = minimize(obj_func, x0=[10, 10], method='SLSQP',constraints=[{'type': 'eq', 'fun': cons_func}], tol = 1e-6)

    L_Phi = -res.fun
    #print( res.x)
    
    return L_Phi



